File: reference/relations.py
"""Section 4.3 of ALGORITHM v4 draft 5.2: identity_relation and friends.
Depends on reference.primitives (types) and reference.engine (DOMAIN, SAT,
decompose/build_gamma for measure-facet disposition), NOT on
reference.extraction (no cycle).
"""

from __future__ import annotations

from typing import Optional, Tuple

from reference.primitives import (
    EXCL_V1,
    Extent,
    Frame,
    MALFORMED_MENTION,
    CONDITION_UNDECIDABLE,
    UNIVERSAL,
    EXISTENTIAL,
    eff_quant,
    worst_cause,
)
from reference.tables import T
from reference import engine

YES, NO, UNKNOWN = "YES", "NO", "UNKNOWN"
COMPARABLE, INERT, UNDECIDABLE = "COMPARABLE", "INERT", "UNDECIDABLE"
MATCH, CONFLICT = "MATCH", "CONFLICT"


def _has_malformed(extent: Extent) -> bool:
    """generalizes step 6: "A carries any malformed value/modifier
    product". Structurally unreachable in this pipeline BY CONSTRUCTION,
    not by fixture coverage: spec section 0's symmetric rule makes every
    malformed value/modifier product abstain() DURING extraction (2.4,
    2.5, 3.2 step 7, 3.2 adjunct rules), which raises before the Extent
    is constructed and drives the FIELD to PARTIAL -- so no Extent
    carrying a malformed product ever reaches identity comparison. The
    step is kept as an explicit named check (returning the invariant's
    value) so the numbered spec steps remain visibly total."""
    return False


def generalizes(A: Extent, B: Extent, same_frame: bool) -> Tuple[str, Optional[str]]:
    if A.facet != B.facet:
        return NO, None
    if not A.subject <= B.subject:
        return NO, None
    if not A.object <= B.object:
        return NO, None

    # step 4: modifiers (spec order: EXCL pair -> NO; no same-rel match
    # in B -> NO; same-rel match(es) exist but none contains a.objset ->
    # UNKNOWN(condition_undecidable). NO is definitive and takes
    # precedence over UNKNOWN when both occur across different
    # modifiers of A.)
    any_unknown = False
    for rel, objset in A.modifiers:
        same_rel_in_b = [ov for r2, ov in B.modifiers if r2 == rel]
        if not same_rel_in_b:
            return NO, None
        for ov in same_rel_in_b:
            for a_term in objset:
                for b_term in ov:
                    if frozenset({a_term, b_term}) in EXCL_V1:
                        return NO, None
        if not any(objset <= ov for ov in same_rel_in_b):
            any_unknown = True
    if any_unknown:
        return UNKNOWN, CONDITION_UNDECIDABLE

    # step 5
    if eff_quant(A.quant) == EXISTENTIAL and not same_frame:
        return UNKNOWN, CONDITION_UNDECIDABLE

    # step 6
    if _has_malformed(A):
        return UNKNOWN, MALFORMED_MENTION

    return YES, None


def identity_relation(A: Extent, D_A, B: Extent, D_B, same_frame: bool) -> str:
    """Returns one of COMPARABLE | INERT | ('UNDECIDABLE', cause). A Rel3
    is NEVER read as boolean; callers branch on all three arms -- this
    reference implementation encodes UNDECIDABLE as the 2-tuple
    ('UNDECIDABLE', cause) so callers cannot accidentally treat it as a
    truthy/falsy COMPARABLE-or-not value."""
    g1, c1 = generalizes(A, B, same_frame)
    g2, c2 = generalizes(B, A, same_frame)
    if g1 != YES and g2 != YES:
        if UNKNOWN in (g1, g2):
            return (UNDECIDABLE, worst_cause([c for c in (c1, c2) if c]))
        return INERT
    d = engine.DOMAIN(None, D_A, D_B)
    if d == "OVERLAP":
        return COMPARABLE
    if d == "DISJOINT":
        return INERT
    # d == UNKNOWN
    cause_ = MALFORMED_MENTION if _domain_has_malformed(D_A, D_B) else CONDITION_UNDECIDABLE
    return (UNDECIDABLE, cause_)


def _domain_has_malformed(D_A, D_B) -> bool:
    c1 = engine.cause(D_A)
    c2 = engine.cause(D_B)
    return MALFORMED_MENTION in (c1, c2)


def rel_is_undecidable(rel) -> bool:
    return isinstance(rel, tuple) and rel and rel[0] == UNDECIDABLE


def rel_cause(rel) -> Optional[str]:
    return rel[1] if rel_is_undecidable(rel) else None


def two_way_generalizes(A: Extent, B: Extent, same_frame: bool) -> Tuple[str, Optional[str]]:
    g1, c1 = generalizes(A, B, same_frame)
    if g1 == YES:
        return YES, None
    g2, c2 = generalizes(B, A, same_frame)
    if g2 == YES:
        return YES, None
    if g1 == NO and g2 == NO:
        return NO, None
    return UNKNOWN, worst_cause([c for c in (c1, c2) if c])


def meet(A: Extent, B: Extent) -> Extent:
    modifiers = A.modifiers if len(A.modifiers) >= len(B.modifiers) else B.modifiers
    q = EXISTENTIAL if (eff_quant(A.quant) == EXISTENTIAL or eff_quant(B.quant) == EXISTENTIAL) else UNIVERSAL
    return Extent(
        facet=A.facet,
        subject=A.subject | B.subject,
        object=A.object | B.object,
        modifiers=modifiers,
        quant=q,
        polarity=A.polarity,
        values=None,
    )


def disposition(a: Frame, b: Frame):
    """MATCH|CONFLICT|('UNDECIDABLE', cause)."""
    facet_a = T.facets_v1.get(a.extent.facet, {})
    is_measure_facet = bool(facet_a.get("measure"))
    if is_measure_facet:
        if a.extent.values is None or b.extent.values is None:
            return (UNDECIDABLE, CONDITION_UNDECIDABLE)
        # spec 2.4: comparisons are legal only within one unit group and
        # one currency -- cross-group/cross-currency -> UnknownAtom with
        # cause malformed_mention, so the disposition is UNDECIDABLE,
        # never a silent MATCH of two independent quantities.
        unit_a = a.extent.values[0].unit
        unit_b = b.extent.values[0].unit
        if unit_a != unit_b:
            return (UNDECIDABLE, MALFORMED_MENTION)
        try:
            compiled = engine.build_varmap(
                [_measure_atom(a.extent), _measure_atom(b.extent)]
            )
        except engine.EnvelopeExceeded:
            return (UNDECIDABLE, CONDITION_UNDECIDABLE)
        conflict = not engine.SAT(
            compiled,
            _and2(_measure_atom(a.extent), _measure_atom(b.extent)),
        )
        return CONFLICT if conflict else MATCH

    if a.extent.polarity != b.extent.polarity and not (
        eff_quant(a.extent.quant) == EXISTENTIAL and eff_quant(b.extent.quant) == EXISTENTIAL
    ):
        return CONFLICT
    return MATCH


def _measure_atom(extent: Extent):
    from reference.primitives import MeasureAtom

    unit = extent.values[0].unit if extent.values else ""
    qty = (extent.facet, extent.subject, unit)
    return MeasureAtom(qty, extent.values or ())


def _and2(a, b):
    from reference.primitives import And

    return And((a, b))

File: reference/ts/src/relations.ts
/**
 * Section 4.3 of ALGORITHM v4 draft 5.4: identity_relation and friends.
 * Depends on primitives.ts (types) and engine.ts (DOMAIN, SAT,
 * decompose/build_gamma for measure-facet disposition), NOT on
 * extraction.ts (no cycle). Mirrors reference/relations.py one-for-one.
 */

import { T, unorderedPairKey } from "./tables.js";
import { FSet } from "./fset.js";
import {
  Bool,
  CONDITION_UNDECIDABLE,
  EXCL_V1,
  EXISTENTIAL,
  Extent,
  MALFORMED_MENTION,
  MeasureAtom,
  MeasureQty,
  UNIVERSAL,
  effQuant,
  mkAnd,
  mkMeasureAtom,
  worstCause,
} from "./primitives.js";
import * as engine from "./engine.js";

export const YES = "YES";
export const NO = "NO";
export const UNKNOWN = "UNKNOWN";
export const COMPARABLE = "COMPARABLE";
export const INERT = "INERT";
export const UNDECIDABLE_TAG = "UNDECIDABLE";
export const MATCH = "MATCH";
export const CONFLICT = "CONFLICT";

export interface UndecidableTag {
  readonly tag: "UNDECIDABLE";
  readonly cause: string;
}

/** Rel3 (TOTAL): COMPARABLE | INERT | UNDECIDABLE(cause). Never read as
 * boolean -- every consumer branches on all three arms via relIsUndecidable
 * / plain string equality against COMPARABLE / INERT. */
export type Rel3 = "COMPARABLE" | "INERT" | UndecidableTag;

export type DispositionResult = "MATCH" | "CONFLICT" | UndecidableTag;

export function relIsUndecidable(r: Rel3 | DispositionResult): r is UndecidableTag {
  return typeof r === "object" && r.tag === "UNDECIDABLE";
}

export function relCause(r: Rel3 | DispositionResult): string | null {
  return relIsUndecidable(r) ? r.cause : null;
}

export type GenResult = readonly ["YES" | "NO" | "UNKNOWN", string | null];

/** generalizes step 6: "A carries any malformed value/modifier product".
 * Structurally unreachable in this pipeline BY CONSTRUCTION, not by
 * fixture coverage: spec section 0's symmetric rule makes every malformed
 * value/modifier product abstain() DURING extraction (2.4, 2.5, 3.2 step
 * 7, 3.2 adjunct rules), which raises before the Extent is constructed and
 * drives the FIELD to PARTIAL -- so no Extent carrying a malformed product
 * ever reaches identity comparison. Kept as an explicit named check
 * (returning the invariant's value) so the numbered spec steps remain
 * visibly total. */
function hasMalformed(_extent: Extent): boolean {
  return false;
}

export function generalizes(A: Extent, B: Extent, sameFrame: boolean): GenResult {
  if (A.facet !== B.facet) return ["NO", null];
  if (!A.subject.isSubsetOf(B.subject)) return ["NO", null];
  if (!A.object.isSubsetOf(B.object)) return ["NO", null];

  // step 4: modifiers (spec order: EXCL pair -> NO; no same-rel match in
  // B -> NO; same-rel match(es) exist but none contains a.objset ->
  // UNKNOWN(condition_undecidable). NO is definitive and takes precedence
  // over UNKNOWN when both occur across different modifiers of A.)
  let anyUnknown = false;
  for (const { rel, objset } of A.modifiers) {
    const sameRelInB: FSet[] = [];
    for (const mb of B.modifiers) {
      if (mb.rel === rel) sameRelInB.push(mb.objset);
    }
    if (sameRelInB.length === 0) return ["NO", null];
    for (const ov of sameRelInB) {
      for (const aTerm of objset.toArray()) {
        for (const bTerm of ov.toArray()) {
          if (EXCL_V1.has(unorderedPairKey(aTerm, bTerm))) {
            return ["NO", null];
          }
        }
      }
    }
    if (!sameRelInB.some((ov) => objset.isSubsetOf(ov))) {
      anyUnknown = true;
    }
  }
  if (anyUnknown) return ["UNKNOWN", CONDITION_UNDECIDABLE];

  // step 5
  if (effQuant(A.quant) === EXISTENTIAL && !sameFrame) {
    return ["UNKNOWN", CONDITION_UNDECIDABLE];
  }

  // step 6
  if (hasMalformed(A)) return ["UNKNOWN", MALFORMED_MENTION];

  return ["YES", null];
}

/** Returns one of COMPARABLE | INERT | UNDECIDABLE(cause). */
export function identityRelation(A: Extent, DA: Bool, B: Extent, DB: Bool, sameFrame: boolean): Rel3 {
  const [g1, c1] = generalizes(A, B, sameFrame);
  const [g2, c2] = generalizes(B, A, sameFrame);
  if (g1 !== "YES" && g2 !== "YES") {
    if (g1 === "UNKNOWN" || g2 === "UNKNOWN") {
      return { tag: "UNDECIDABLE", cause: worstCause([c1, c2]) };
    }
    return "INERT";
  }
  const d = engine.DOMAIN(null, DA, DB);
  if (d === "OVERLAP") return "COMPARABLE";
  if (d === "DISJOINT") return "INERT";
  // d == UNKNOWN
  const causeVal = domainHasMalformed(DA, DB) ? MALFORMED_MENTION : CONDITION_UNDECIDABLE;
  return { tag: "UNDECIDABLE", cause: causeVal };
}

function domainHasMalformed(DA: Bool, DB: Bool): boolean {
  const c1 = engine.causeOf(DA);
  const c2 = engine.causeOf(DB);
  return c1 === MALFORMED_MENTION || c2 === MALFORMED_MENTION;
}

export function twoWayGeneralizes(A: Extent, B: Extent, sameFrame: boolean): GenResult {
  const [g1, c1] = generalizes(A, B, sameFrame);
  if (g1 === "YES") return ["YES", null];
  const [g2, c2] = generalizes(B, A, sameFrame);
  if (g2 === "YES") return ["YES", null];
  if (g1 === "NO" && g2 === "NO") return ["NO", null];
  return ["UNKNOWN", worstCause([c1, c2])];
}

export function meet(A: Extent, B: Extent): Extent {
  const modifiers = A.modifiers.size >= B.modifiers.size ? A.modifiers : B.modifiers;
  const q = effQuant(A.quant) === EXISTENTIAL || effQuant(B.quant) === EXISTENTIAL ? EXISTENTIAL : UNIVERSAL;
  return {
    facet: A.facet,
    subject: A.subject.union(B.subject),
    object: A.object.union(B.object),
    modifiers,
    quant: q,
    polarity: A.polarity,
    values: null,
  };
}

export interface HasExtent {
  readonly extent: Extent;
}

export function measureAtomOf(extent: Extent): MeasureAtom {
  const unit = extent.values && extent.values.length > 0 ? extent.values[0]!.unit : "";
  const qty: MeasureQty = [extent.facet, extent.subject, unit];
  return mkMeasureAtom(qty, extent.values ?? []);
}

/** MATCH|CONFLICT|UNDECIDABLE(cause). */
export function disposition(a: HasExtent, b: HasExtent): DispositionResult {
  const facetA = T.facetsV1[a.extent.facet];
  const isMeasureFacet = Boolean(facetA?.measure);
  if (isMeasureFacet) {
    if (a.extent.values === null || b.extent.values === null) {
      return { tag: "UNDECIDABLE", cause: CONDITION_UNDECIDABLE };
    }
    // spec 2.4: comparisons are legal only within one unit group and one
    // currency -- cross-group/cross-currency -> UnknownAtom with cause
    // malformed_mention, so the disposition is UNDECIDABLE, never a
    // silent MATCH of two independent quantities.
    const unitA = a.extent.values[0]!.unit;
    const unitB = b.extent.values[0]!.unit;
    if (unitA !== unitB) {
      return { tag: "UNDECIDABLE", cause: MALFORMED_MENTION };
    }
    let compiled: engine.Compiled;
    try {
      compiled = engine.buildVarmap([measureAtomOf(a.extent), measureAtomOf(b.extent)]);
    } catch (e) {
      if (e instanceof engine.EnvelopeExceeded) {
        return { tag: "UNDECIDABLE", cause: CONDITION_UNDECIDABLE };
      }
      throw e;
    }
    const conflict = !engine.SAT(compiled, mkAnd([measureAtomOf(a.extent), measureAtomOf(b.extent)]));
    return conflict ? "CONFLICT" : "MATCH";
  }

  if (
    a.extent.polarity !== b.extent.polarity &&
    !(effQuant(a.extent.quant) === EXISTENTIAL && effQuant(b.extent.quant) === EXISTENTIAL)
  ) {
    return "CONFLICT";
  }
  return "MATCH";
}
